### PR TITLE
Split Timestamp Class Into Time, Timestamp,  and Duration

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Split the `Timestamp` class into a hierarchy, with a `Time` class being the parent of both a `Timestamp` and `Duration` class.

### Testing Done
Unit tests for new classes and changed functionality.

### Length Justification
Timestamps are in a lot of places.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/world/team.cpp [hunk 1]
```diff
@@ -61,6 +61,12 @@ void Team::removeExpiredRobots(const Timestamp& timestamp)
     // has passed, then remove the robot from the team
     for (auto it = team_robots.begin(); it != team_robots.end();)
     {
+        Duration time_diff = timestamp - it->second.lastUpdateTimestamp();
+        if (time_diff.getSeconds() < 0)
+        {
+            throw std::invalid_argument(
+                "Error: tried to remove a robot at a negative time");
+        }
         if ((timestamp - it->second.lastUpdateTimestamp()) > robot_expiry_buffer_duration)
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/world/team.cpp]: Can we replace `(timestamp - it->second.lastUpdateTimestamp())` with the new `time_diff` here?

File: src/thunderbots/software/network_input/filter/ball_filter.h [hunk 2]
```diff
@@ -4,20 +4,20 @@
...
+#include "util/time/timestamp.h"
 
 /**
  * A lightweight datatype used to input new data into the filter.
  * We do this rather than taking the SSL_DetectionBall directly
  * so we can make this module more generic and abstract away
  * the protobuf for testing
  */
-typedef struct
+struct SSLBallDetection
```
Comment (MathewMacDougall) [inline src/thunderbots/software/network_input/filter/ball_filter.h:15]: Why remove the typedef?

File: src/thunderbots/software/network_input/filter/robot_filter.h [hunk 3]
```diff
@@ -4,22 +4,22 @@
...
+#include "util/time/timestamp.h"
 
 /**
  * A lightweight datatype used to input new data into the filter.
  * We do this rather than taking the SSL_Detection data directly
  * so we can make this module more generic and abstract away
  * the protobuf for testing
  */
-typedef struct SSLRobotData_t
+struct SSLRobotDetection
```
Comment (MathewMacDougall) [inline src/thunderbots/software/network_input/filter/robot_filter.h:15]: Why remove the typedef?

File: src/thunderbots/software/test/util/time/duration.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    EXPECT_DOUBLE_EQ(duration.getSeconds(), -0.2);
+}
+
+TEST(DurationTest, create_duration_from_zero_milliseconds)
+{
+    Duration duration = Duration::fromMilliseconds(0);
+    EXPECT_DOUBLE_EQ(duration.getMilliseconds(), 0);
+}
+
+TEST(DurationTest, create_duration_from_negative_milliseconds)
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/util/time/duration.cpp:31]: Missing test case for `create_duration_from_positive_milliseconds`?

File: src/thunderbots/software/test/util/time/duration.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+{
+    Duration duration = Duration::fromMilliseconds(-125);
+    EXPECT_DOUBLE_EQ(duration.getMilliseconds(), -125);
+}
+
+TEST(DurationTest, get_seconds_duration_in_milliseconds)
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/util/time/duration.cpp:43]: Missing test case for `get_seconds_duration_in_seconds`?

File: src/thunderbots/software/test/util/time/time.cpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    EXPECT_DOUBLE_EQ(time.getSeconds(), -0.2);
+}
+
+TEST(TimeTest, create_time_from_zero_milliseconds)
+{
+    Time time = Time::fromMilliseconds(0);
+    EXPECT_DOUBLE_EQ(time.getMilliseconds(), 0);
+}
+
+TEST(TimeTest, create_time_from_negative_milliseconds)
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/util/time/time.cpp]: Missing test case for `create_time_from_positive_milliseconds`?

File: src/thunderbots/software/util/time/duration.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    Duration operator-(const Duration& duration) const;
+
+   private:
+    /**
+     * Constructs a Duration value from a value in seconds.
+     * @param timestamp_seconds A value >= 0.0, in seconds, from which to create the
+     * timestamp
+     */
+    explicit Duration(double timestamp_seconds);
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/duration.h]:
```suggestion
    explicit Duration(double duration_seconds);
```

File: src/thunderbots/software/util/time/duration.cpp [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * Implementation for the Duration class
+ */
+
+#include "duration.h"
+
+#include "shared/constants.h"
+
+Duration::Duration() : Duration(0) {}
+
+Duration::Duration(double timestamp_seconds) : Time(timestamp_seconds) {}
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/duration.cpp]:
```suggestion
Duration::Duration(double duration_seconds) : Time(duration_seconds) {}
```

File: src/thunderbots/software/util/time/time.h [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+     */
+    Time operator-(const Time& time) const;
+
+   protected:
+    /**
+     * Constructs a Time value from a value in seconds.
+     * @param Time_seconds A value in seconds, from which to create the
+     * Time
+     */
+    explicit Time(double Time_seconds);
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/time.h]:
```suggestion
    explicit Time(double time_seconds);
```

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Nice work, this looks really clean. Mostly small comments for typos and test cases

File: src/thunderbots/software/test/util/time/time.cpp (hunk 6 above)
Comment (MathewMacDougall) [inline src/thunderbots/software/test/util/time/time.cpp]: Missing test case for `create_time_from_positive_milliseconds`?

File: src/thunderbots/software/util/time/duration.cpp (hunk 8 above)
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/duration.cpp]:
```suggestion
Duration::Duration(double duration_seconds) : Time(duration_seconds) {}
```

File: src/thunderbots/software/util/time/duration.h (hunk 7 above)
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/duration.h]:
```suggestion
    explicit Duration(double duration_seconds);
```

File: src/thunderbots/software/util/time/time.h (hunk 9 above)
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/time.h]:
```suggestion
    explicit Time(double time_seconds);
```

File: src/thunderbots/software/util/time/timestamp.cpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+const Timestamp Timestamp::fromMilliseconds(double milliseconds)
+{
+    return Timestamp(milliseconds * SECONDS_PER_MILLISECOND);
+}
+
+Timestamp Timestamp::operator+(const Duration &duration) const
+{
+    // TODO
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/time/timestamp.cpp]: Is this done?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Nice work, this looks clean.

Do we not want the `Time` class to be pure-virtual? I didn't think `Time` would be its own instantiable class, since we would have the `Timestamp` and `Duration` implementations.

File: src/thunderbots/software/CMakeLists.txt [hunk 11]
```diff
@@ -443,12 +452,18 @@ if (CATKIN_ENABLE_TESTING)
...
 
-    catkin_add_gtest(timestamp_test
-            test/util/timestamp.cpp
-            util/timestamp.cpp
+    catkin_add_gtest(time_test
+            test/util/time/main.cpp
+            test/util/time/timestamp.cpp
+            test/util/time/time.cpp
+            test/util/time/duration.cpp
+            util/time/time.cpp
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/CMakeLists.txt]: Removed two instances

File: src/thunderbots/software/ai/world/team.cpp (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/world/team.cpp]: absolutely, changed

File: src/thunderbots/software/network_input/filter/ball_filter.h (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/network_input/filter/ball_filter.h:15]: I didn't see any reason for it, and removing it simplified things a bit (just in terms of syntax, this is semantically the same). Just a little cleanup

File: src/thunderbots/software/network_input/filter/robot_filter.h (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/network_input/filter/robot_filter.h:15]: See comment for `SSLBallDetection`

File: src/thunderbots/software/test/util/time/duration.cpp (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/test/util/time/duration.cpp:31]: Added

File: src/thunderbots/software/test/util/time/duration.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/test/util/time/duration.cpp:43]: added

File: src/thunderbots/software/test/util/time/time.cpp (hunk 6 above)
Comment (garethellis0) [reply src/thunderbots/software/test/util/time/time.cpp]: Holding off on this until we decide what do do about the `Time` class

Comment (garethellis0) [reply src/thunderbots/software/test/util/time/time.cpp]: Added

File: src/thunderbots/software/util/time/duration.h (hunk 7 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/duration.h]: changed

File: src/thunderbots/software/util/time/duration.cpp (hunk 8 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/duration.cpp]: changed

File: src/thunderbots/software/util/time/time.h (hunk 9 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/time.h]: Changed

File: src/thunderbots/software/util/time/timestamp.cpp (hunk 10 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/timestamp.cpp]: whoops, yup, removed.

File: src/thunderbots/software/util/time/time.h (hunk 9 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/time.h]: changed

File: src/thunderbots/software/util/time/duration.h (hunk 7 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/duration.h]: changed

File: src/thunderbots/software/util/time/duration.cpp (hunk 8 above)
Comment (garethellis0) [reply src/thunderbots/software/util/time/duration.cpp]: changed

File: src/thunderbots/software/test/util/time/time.cpp (hunk 6 above)
Comment (garethellis0) [reply src/thunderbots/software/test/util/time/time.cpp]: added

**Comment by garethellis0** [comment]
Moving discussion about making `Time` virtual to Slack.

**Comment by garethellis0** [comment]
After some discussion, made `Time` virtual.

File: src/thunderbots/software/network_input/filter/ball_filter.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/filter/ball_filter.h:15]: But doesn't the typedef allow variables to store the struct instance without the `struct` prefix?

https://stackoverflow.com/questions/1675351/typedef-struct-vs-struct-definitions

**Comment by MathewMacDougall** [comment]
Just one last follow up regarding the struct typdefs, otherwise LGTM.

File: src/thunderbots/software/network_input/filter/ball_filter.h (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/network_input/filter/ball_filter.h:15]: Close, but not the case in C++. If you scroll down a bit in that link you'll see this example:
```
 // C++
struct S { 
    int x; 
}; // S defined as a class

void f( S a ); // correct: struct is optional
```

**Review by MathewMacDougall — APPROVED** [review]

**Comment by MathewMacDougall** [comment]
LGTM :+1:

**Pull request merged**